rpcs3/Emu/RSX/Common/BufferUtils.cpp

```diff
@@ -755,39 +762,290 @@ namespace
+			for (unsigned n = 0; n < iterations; ++n)
+			{
+				const __m256i raw = _mm256_loadu_si256(src_stream++);
+				const __m256i value = _mm256_shuffle_epi8(raw, mask);
+				const __m256i mask = _mm256_cmpeq_epi16(restart, value);
+				const __m256i tmp = _mm256_andnot_si256(mask, value);
+				max = _mm256_max_epu16(max, tmp);
+				min = _mm256_min_epu16(min, value);
+				_mm256_storeu_si256(dst_stream++, value);
+			}
+
+			const __m256i mask_step1 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0x1F, 0x1E, 0x1D, 0x1C, 0x1B, 0x1A, 0x19, 0x18,
+				0x17, 0x16, 0x15, 0x14, 0x13, 0x12, 0x11, 0x10);
+
+			const __m256i mask_step2 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0xF, 0xE, 0xD, 0xC, 0xB, 0xA, 0x9, 0x8);
+
+			const __m256i mask_step3 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0x7, 0x6, 0x5, 0x4);
+
+			const __m256i mask_step4 = _mm256_set_epi8(
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0x3, 0x2);
+
+			__m256i tmp = _mm256_shuffle_epi8(min, mask_step1);
+			min = _mm256_min_epu16(min, tmp);
+			tmp = _mm256_shuffle_epi8(min, mask_step2);
+			min = _mm256_min_epu16(min, tmp);
+			tmp = _mm256_shuffle_epi8(min, mask_step3);
+			min = _mm256_min_epu16(min, tmp);
+			tmp = _mm256_shuffle_epi8(min, mask_step4);
+			min = _mm256_min_epu16(min, tmp);
+
+			tmp = _mm256_shuffle_epi8(max, mask_step1);
+			max = _mm256_max_epu16(max, tmp);
+			tmp = _mm256_shuffle_epi8(max, mask_step2);
+			max = _mm256_max_epu16(max, tmp);
+			tmp = _mm256_shuffle_epi8(max, mask_step3);
+			max = _mm256_max_epu16(max, tmp);
+			tmp = _mm256_shuffle_epi8(max, mask_step4);
+			max = _mm256_max_epu16(max, tmp);
+
+			const u16 min_index = u16(_mm_cvtsi128_si32(_mm256_castsi256_si128(min)) & 0xFFFF);
+			const u16 max_index = u16(_mm_cvtsi128_si32(_mm256_castsi256_si128(max)) & 0xFFFF);
+
+			return std::make_tuple(min_index, max_index, count);
+		}
+
+		SSE4_1_FUNC
 		static
-		std::tuple<T, T, u32> upload_untouched(gsl::span<to_be_t<const T>> src, gsl::span<T> dst, u32 restart_index, bool skip_restart)
+		std::tuple<u16, u16, u32> upload_u16_swapped_sse4_1(const void *src, void *dst, u32 count, u16 restart_index)
 		{
-			T min_index = index_limit<T>(), max_index = 0;
 			u32 dst_index = 0;
 
-			for (const T index : src)
+			const __m128i mask = _mm_set_epi8(
+				0xE, 0xF, 0xC, 0xD,
+				0xA, 0xB, 0x8, 0x9,
+				0x6, 0x7, 0x4, 0x5,
+				0x2, 0x3, 0x0, 0x1);
+
+			auto src_stream = (const __m128i*)src;
+			auto dst_stream = (__m128i*)dst;
+
+			__m128i restart = _mm_set1_epi16(restart_index);
+			__m128i min = _mm_set1_epi16(0xffff);
+			__m128i max = _mm_set1_epi16(0);
+
+			const auto iterations = count / 8;
+			for (unsigned n = 0; n < iterations; ++n)
 			{
+				const __m128i raw = _mm_loadu_si128(src_stream++);
```

Comment:
There is a flaw in the logic here; let me explain.
1. The PS3 (and older GPU APIs for that matter, including OpenGL) allow setting anything as the restart index. You can use 0, 1, 0xffff, 0xdead, whatever you like. However...
DX12 and Vulkan do not allow this. For this reason, we have to replace all instances of primitive restart with UINT_MAX for the type, i.e ffff for u16 and ffffffff for u32.
2. The index range returned [min_index, max_index] should not take the primitive restart index into account. It defines a memory access range and the primitive restart index is almost supposed to be outside this range.

So here comes the huge problem; horizontal min/max while ignoring the primitive restart index. Its not impossible just needs more code.
The simplified logic is:

value = shuffle(raw)
mask = cmpeq(restart, value)
tmp = andnot(mask, value)  // 0s where value == index
final = tmp | mask                // All restarts pushed to uint_max
max = max(max, tmp)
min = min(min, final)
write(final)

Reply:
This comment applies to all the variations provided.

Reply:
This is now fixed, thanks.

It wasn’t an issue in NieR because it already uses `0xffff` for its restart index.